error: endpoint not found

**Describe the bug**
upgrading from version .150 to .179 breaks majority of routes, Skipper returns "no endpoint" when accessing these routes

**To Reproduce**

errors logged
```
[APP]time="2020-11-02T07:46:29Z" level=error msg="convertPathRule: add shuntroute to return 502 for ingress status/status service status with 0 endpoints: endpoint not found"
```

Slack ref: https://gophers.slack.com/archives/C82Q5JNH5/p1604304999030600

@szuecs

I can pinpoint the breaking change was introduced in version .174.
Actually with that version I even get a nil-pointer panic at this critical point!
```
msg="route settings, reset, route: kube_monitoring__alertmanager_main__monitoring_org___alertmanager__alertmanager_main:  │
│ %!v(PANIC=String method: runtime error: invalid memory address or nil pointer dereference)"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x788366]
goroutine 49 [running]:
github.com/zalando/skipper/routing.createFilter(0xc000285860, 0x0, 0xc000342d50, 0x2, 0x3, 0xc00017e2e0, 0x2)
    /workspace/routing/datasource.go:221 +0x26
github.com/zalando/skipper/routing.createFilters(0xc000285860, 0xc0003421b0, 0x5, 0x6, 0xc000342d50, 0x0, 0x0, 0x0, 0x0, 0xc0002dfb10)
    /workspace/routing/datasource.go:242 +0x9f
github.com/zalando/skipper/routing.processRouteDef(0xc000342d50, 0xc000285860, 0xc000338a20, 0x1, 0x2, 0xc0003221c0)
    /workspace/routing/datasource.go:465 +0xd0
github.com/zalando/skipper/routing.processRouteDefs(0xc000285860, 0x0, 0xb2d05e00, 0xc00027eb70, 0x1, 0x1, 0xc000260900, 0x11, 0x12, 0x20, ...)
    /workspace/routing/datasource.go:502 +0x105
github.com/zalando/skipper/routing.receiveRouteMatcher(0xc000285860, 0x0, 0xb2d05e00, 0xc00027eb70, 0x1, 0x1, 0xc000260900, 0x11, 0x12, 0x20, ...)
    /workspace/routing/datasource.go:539 +0x345
created by github.com/zalando/skipper/routing.(*Routing).startReceivingUpdates
    /workspace/routing/routing.go:334 +0xa9
```

--- a/dataclients/kubernetes/ingress_test.go
+++ b/dataclients/kubernetes/ingress_test.go
@@ -1,0 +1,11 @@
+package kubernetes_test
+
+import (
+	"testing"
+
+	"github.com/zalando/skipper/dataclients/kubernetes/kubernetestest"
+)
+
+func TestIngressFixtures(t *testing.T) {
+	kubernetestest.FixturesToTest(t, "testdata/ingress/named-ports")
+}

--- a/dataclients/kubernetes/ingressdefinitions.go
+++ b/dataclients/kubernetes/ingressdefinitions.go
@@ -71,6 +71,7 @@ func formatEndpoint(a *address, p *port, protocol string) string {
 	return fmt.Sprintf("%s://%s:%d", protocol, a.IP, p.Port)
 }
 
+// TODO(sszuecs): refactoring and replace it with clusterState.getEndpointsByTarget, but we need to support protocol to have feature parity with current ingress
 // svcPortName is the truncated value of int or string from kubernetes service svcPort (should not be used here)
 // svcPortTarget is the truncated value of int or string from kubernetes service targetPort
 func (ep endpoint) targets(svcPortName, svcPortTarget, protocol string) []string {

--- a/dataclients/kubernetes/ingressdefinitions.go
+++ b/dataclients/kubernetes/ingressdefinitions.go
@@ -71,6 +71,8 @@ func formatEndpoint(a *address, p *port, protocol string) string {
 	return fmt.Sprintf("%s://%s:%d", protocol, a.IP, p.Port)
 }
 
+// svcPortName is the truncated value of int or string from kubernetes service svcPort (should not be used here)
+// svcPortTarget is the truncated value of int or string from kubernetes service targetPort
 func (ep endpoint) targets(svcPortName, svcPortTarget, protocol string) []string {
 	result := make([]string, 0)
 	for _, s := range ep.Subsets {
@@ -82,7 +84,7 @@ func (ep endpoint) targets(svcPortName, svcPortTarget, protocol string) []string
 			//
 			// https://github.com/zalando/skipper/tree/improvement/service-port-fallback-handling
 			//
-			if port.Name == svcPortName || strconv.Itoa(port.Port) == svcPortTarget {
+			if port.Name == svcPortName || port.Name == svcPortTarget || strconv.Itoa(port.Port) == svcPortTarget {
 				for _, addr := range s.Addresses {
 					result = append(result, formatEndpoint(addr, port, protocol))
 				}

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport.yaml
@@ -1,0 +1,49 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    targetPort: web
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: foo
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport2.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport2.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport2.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport2.yaml
@@ -1,0 +1,49 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: web
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    targetPort: web
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: foo
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport3.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport3.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport3.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport3.yaml
@@ -1,0 +1,52 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    targetPort: web
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: foo
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+  - name: web
+    port: 80
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport4.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport4.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport4.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport4.yaml
@@ -1,0 +1,52 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: web
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    targetPort: web
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: foo
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    #  - name: web
+    #    port: 80
+    #    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport5.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport5.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport5.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-multiport5.yaml
@@ -1,0 +1,52 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: web
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    targetPort: foo
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: web
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8081
+    protocol: TCP
+  - name: web
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-no-match.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-no-match.eskip
@@ -1,0 +1,5 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> status(502)
+	-> inlineContent("no endpoints")
+	-> <shunt>;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-no-match.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-no-match.yaml
@@ -1,0 +1,45 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: web
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 80
+    protocol: TCP
+    targetPort: foo
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-same-name.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-same-name.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-same-name.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports-same-name.yaml
@@ -1,0 +1,45 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: web
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: web
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports.yaml
@@ -24,7 +24,7 @@ spec:
   - name: web
     port: 80
     protocol: TCP
-    targetPort: web
+    targetPort: foo
   selector:
     application: myapp
   type: ClusterIP
@@ -40,6 +40,6 @@ subsets:
   - ip: 10.2.9.103
   - ip: 10.2.9.104
   ports:
-  - name: web
+  - name: foo
     port: 8080
     protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/named-ports.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/named-ports.yaml
@@ -1,0 +1,45 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: web
+    port: 80
+    protocol: TCP
+    targetPort: web
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: web
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/numbered-ports-no-match.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/numbered-ports-no-match.eskip
@@ -1,0 +1,5 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> status(502)
+	-> inlineContent("no endpoints")
+	-> <shunt>;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/numbered-ports-no-match.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/numbered-ports-no-match.yaml
@@ -1,0 +1,45 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP
+    targetPort: foo
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - name: foo
+    port: 8080
+    protocol: TCP

--- a/dataclients/kubernetes/testdata/ingress/named-ports/ports-without-name.eskip
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/ports-without-name.eskip
@@ -1,0 +1,3 @@
+kube_default__myapp__example_org____myapp:
+	Host("^example[.]org$")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;

--- a/dataclients/kubernetes/testdata/ingress/named-ports/ports-without-name.yaml
+++ b/dataclients/kubernetes/testdata/ingress/named-ports/ports-without-name.yaml
@@ -1,0 +1,43 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: myapp
+  namespace: default
+spec:
+  rules:
+  - host: example.org
+    http:
+      paths:
+      - backend:
+          serviceName: myapp
+          servicePort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+spec:
+  clusterIP: 10.3.190.97
+  ports:
+  - port: 80
+    protocol: TCP
+    targetPort: 8080
+  selector:
+    application: myapp
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  name: myapp
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
+  ports:
+  - protocol: TCP
+    port: 8080

